### PR TITLE
[Runtime] : Fix bundling a standard Symfony app into a PHAR

### LINKED ISSUE #49443: [Runtime] Not compatible with PHARs
### Symfony version(s) affected

6.2.6

### Description

This is a follow up of #41771, which already reported the problem. The solution offered there was to use an alternative runtime template but I think the Symfony one should work out of the box with PHARs.

Related Box issues:
- https://github.com/box-project/box/issues/655
- https://github.com/box-project/box/pull/881
- https://github.com/box-project/box/pull/886

Bundling a standard Symfony app into a PHAR using the default runtime component results in:

```
PHP Warning:  Constant  already defined in /home/imper/projects/phartest/build/app.phar on line 16
PHP Fatal error:  Uncaught TypeError: Invalid return value: callable object expected, "int" returned from "./build/app.phar". in phar:///home/imper/projects/phartest/build/app.phar/vendor/autoload_runtime.php:12
Stack trace:
#0 phar:///home/imper/projects/phartest/build/app.phar/bin/console(10): require_once()
#1 /home/imper/projects/phartest/build/app.phar(14): require('...')
#2 {main}
  thrown in phar:///home/imper/projects/phartest/build/app.phar/vendor/autoload_runtime.php on line 12
```

### How to reproduce

Checkout https://github.com/box-project/box/pull/886 and execute `make e2e_symfony_runtime`.

Alternatively:

- Create a brand new standard Symfony app
- Get your hand on a [Box](https://github.com/box-project/box) binary
- Create the following Box config:

```json
{
    "files-bin": [
        ".env.local.php",
        "src/Controller/.gitignore"
    ],
    "directories": [
        "config",
        "var"
    ],
    "force-autodiscovery": true,
    "check-requirements": false,
    "exclude-composer-files": false
}
```

The TL:DR; from this config is: nothing really fancy, include somehow the Symfony files that cannot be inferred by Composer.

And then, run:

```
$ box compile
$ ./bin/console.phar
```

### Possible Solution

The solution suggested in #41771 is in my opinion not satisfactory. Indeed:

- this requires a number of non-trivial steps in order to make it work
- creating a new runtime template which actually no longer really fits the runtime component. As a consequence you do end up with something working for the PHAR, but outside the PHAR, you console script is no longer working with the runtime context. So you might as well just ditch the runtime component there IMO, it is simpler to fallback to the old console script.

I have checked the script and the issue is that somehow:

```
// autoload_runtime.php
$app = require $_SERVER['SCRIPT_FILENAME'];
```

ends up with `$app = 1`, but I just don't get why, it requires the file correctly and the return value should be the callable (even when doing step by step debugging, you end up doing the right path). It's as if the callable was suddenly magically mapped to 1 in the context of the PHAR.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by theofidry**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #49443
| License       | MIT
| Doc PR        | none

This PR proposes to add a (computationally) cheap extension point to edit the application entry point (previously hardcoded `$_SERVER['SCRIPT_FILENAME']`). Indeed in some contexts, for example PHARs, the entry point may change and offering this extension points allows to more easily edit the template: instead of having to register a brand new template (which changes only the required entry point file path), you can now instead export a variable.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I don't think we should go this way. This is on the hot path and Phar is an optional extension. This template file is here to be replaced when the use case needs it. Box is one of them to me.

**Comment by theofidry** [comment]
>This is on the hot path

Regarding the hotpath: would an `extension_loaded` or `Phar::isRunning()` really be a problem? IMO if performance is _such_ a concern that even this kind of if statement is problematic, one would use their own template to avoid the subsequent `is_object` check as well.

> and Phar is an optional extension

Is it really? I thought it was bundled in the PHP binary, I never managed to disable that extension in the end. But this can be easily be fixed with an `extension_loaded` check

>This template file is here to be replaced when the use case needs it. Box is one of them to me.

Note that this is not Box specific: this is PHAR bundler specific, Box or others.

File: src/Symfony/Component/Runtime/Internal/autoload_runtime.template [hunk 1]
```diff
@@ -2,16 +2,30 @@
...
 // autoload_runtime.php @generated by Symfony Runtime
 
-if (true === (require_once __DIR__.'/autoload.php') || empty($_SERVER['SCRIPT_FILENAME'])) {
+$scriptFileName = $_SERVER['SCRIPT_FILENAME'] ?? null;
+
+if (true === (require_once __DIR__.'/autoload.php') || null === $scriptFileName || '' === $scriptFileName) {
     return;
 }
 
+if (\extension_loaded('phar') {
```
Comment (wouterj) [inline src/Symfony/Component/Runtime/Internal/autoload_runtime.template]:
```suggestion
if (\extension_loaded('phar')) {
```

**Comment by theofidry** [comment]
Just a clarification regarding the hotpath: I think it's worth talking of the hotpath _outside_ of the PHAR context, so here it will only be `extension_loaded` and `Phar::running()` calls, which I think are [_pretty fast_](https://github.com/php/php-src/blob/master/ext/phar/phar_object.c#L406).

Once inside a PHAR, I think any extra code is no longer a hot-path: the performance within a PHAR are always widely different due to how PHAR are working.

**Comment by theofidry** [comment]
Another alternative is to go with the other proposed solution to allow a different entry point than `$_SERVER['SCRIPT_FILENAME']`, e.g. `$_SERVER['APP_SCRIPT_FILENAME']`. If the default runtime template accepts this then, although it won't work out of the box still, at the very least it only requires to update the `bin/console` file a bit:

```php
<?php

// autoload_runtime.php @generated by Symfony Runtime

$scriptFileName = $_SERVER['APP_SCRIPT_FILENAME'] ?? $_SERVER['SCRIPT_FILENAME'] ?? null;

// unchanged; just use $scriptFileName instead of  $_SERVER['SCRIPT_FILENAME'];
```

Which from the hotpath perspective only adds the null-coalesce.

**Comment by theofidry** [comment]
Updated the PR to show the code; but I suppose this would be considered a new feature

**Comment by theofidry** [comment]
I updated the PR description as well

**Comment by theofidry** [comment]
Closing, we found a nicer solution 😄

**Comment by tacman** [comment]
> Closing, we found a nicer solution 😄

@theofidry Can you share that solution?   I'm getting a Runtime error that i suspect your solution will solve.  Thx.

```
PHP Fatal error:  Cannot redeclare interface Symfony\Component\Runtime\RunnerInterface (previously declared in /home/tac/g/sites/phar-demo/.box_dump/vendor/symfony/runtime/RunnerInterface.php:17) in /home/tac/g/sites/phar-demo/vendor/symfony/runtime/RunnerInterface.php on line 17
Symfony\Component\ErrorHandler\Error\FatalError^ {#68
  #message: "Compile Error: Cannot redeclare interface Symfony\Component\Runtime\RunnerInterface (previously declared in /home/tac/g/sites/phar-demo/.box_dump/vendor/symfony/runtime/RunnerInterface.php:17)"
  #code: 0
  #file: "/home/tac/g/sites/phar-demo/vendor/symfony/runtime/RunnerInterface.php"
  #line: 17
  -error: array:4 [
    "type" => 64
    "message" => "Cannot redeclare interface Symfony\Component\Runtime\RunnerInterface (previously declared in /home/tac/g/sites/phar-demo/.box_dump/vendor/symfony/runtime/RunnerInterface.php:17)"
    "file" => "/home/tac/g/sites/phar-demo/vendor/symfony/runtime/RunnerInterface.php"
    "line" => 17
```

**Comment by theofidry** [comment]
@tacman yes see https://github.com/symfony/symfony/issues/49443#issuecomment-1847349399

**Comment by tacman** [comment]
Thanks, I saw that, but wasn't sure where to put it.  That is, am I suppose to put it somewhere?  It's an auto-generated file, right?

Are there any repos with a command that compiles to a phar with box I could copy from?

**Pull request closed without merging**